File: ABC087B.ml
(* ABC087B - Coins *)

let solver a b c x =
  let counter = ref 0 in
  for a' = 0 to a do
    for b' = 0 to b do
      for c' = 0 to c do
        if (500 * a') + (100 * b') + (50 * c') = x then counter := !counter + 1
      done
    done
  done;
  !counter

let () =
  let a = read_int () in
  let b = read_int () in
  let c = read_int () in
  let x = read_int () in
  solver a b c x |> string_of_int |> print_endline

File: ABC087B.cpp
// ABC087B - Coins
#include <iostream>

long solver(long A, long B, long C, long X)
{
    auto counter = 0;
    for (auto a = 0; a <= A; ++a)
    {
        for (auto b = 0; b <= B; ++b)
        {
            for (auto c = 0; c <= C; ++c)
            {
                if ((500 * a) + (100 * b) + (50 * c) == X)
                {
                    ++counter;
                }
            }
        }
    }
    return counter;
}

int main()
{
    long A, B, C, X;
    std::cin >> A >> B >> C >> X;
    std::cout << solver(A, B, C, X) << std::endl;
    return 0;
}
